docs/proposals/clusters-in-any-namespace.md

```diff
@@ -0,0 +1,187 @@
+  name: test                              
+  namespace: user-tenant-A        
+type: Opaque                   
+data:
+  server: https://kubernetes-api-server.com:4663
+  project: sample
+  namespaces: dev, stage
+
+----
+kind: Secret
+metadata:
+  labels:
+    argocd.argoproj.io/secret-type: cluster
+  name: test                              
+  namespace: user-tenant-B     
+type: Opaque                   
+data:
+  server: https://kubernetes-api-server.com:4663
+  project: sample
+  namespaces: stage, prod
+
+----
+
+apiVersion: argoproj.io/v1alpha1
+kind: AppProject
+metadata:
+  name: sample
+  namespace: argocd
+spec:
+  permitOnlyProjectScopedClusters: true
+```
+
+### Security Considerations
+
+* How does this proposal impact the security aspects of Argo CD workloads ?
+* Are there any unresolved follow-ups that need to be done to make the enhancement more robust ?
+
+### Risks and Mitigations
+
+1. An attacker or a bot might consider a DDOS attack to overwhelm Argo CD by creating a large number of cluster secrets in different namespaces. To mitigate this problem, Argo CD should only reconcile cluster secrets that are created by trusted tenants. We reconcile secrets only from the namespaces specified in the `--application-namespaces` flag. Since this flag is usually set by an Argo CD admin, we can assume a certain level of trust in these namespaces. Even if an attacker has access to one of these valid namespaces the admin could easily disable these namespaces by updating the flag.
+2. A user might try deploying to a destination that is prohibited. Every cluster secret created by a tenant is scoped to an AppProject that is managed by an admin. The admin could configure the AppProject rules to prevent users from accessing a prohibited destination.
+
+### Upgrade / Downgrade Strategy
+
+Existing Argo CD users should not be affected during upgrade/downgrade as long as they have clusters only in the control-plane namespace. Once they upgrade to a version with this feature, an admin can enable this feature by specifying the `--application-namespaces` flag. If the flag was already present because of Applications in any namespace feature, Argo CD starts reconciling cluster secrets from these defined namespaces.
+
+Downgrading will affect the users because Argo CD will no longer reconcile the clusters from a non-control-plane namespace. Applications using these clusters might fail because Argo CD doesn't recognize the cluster anymore. Users must make sure to migrate cluster secrets from different namespaces to a control-plane namespace before downgrading to a previous version.
+
+## Open Questions [optional]
+
+1. How should we handle multiple cluster secrets pointing to the same URL?
+
+## Drawbacks
+
+1. The Cluster cache needs to be updated to support cluster secrets with the same URL.
+2. Downgrading cannot be easily achieved.
+  
+## Alternatives
+
+Continue with the present approach of managing cluster secrets in a control-plane namespace.
```

Comment:
Why not deploy multiple instances of Argo CD in the same cluster. One instance for each tenant. If the goal is to have the same endpoint, perhaps the baseUrl of the instances could be
- https://argo.my.company.com/tenant-1
- https://argo.my.company.com/tenant-2

This pattern would truly create multi-tenancy at a very low cost.

Reply:
This proposal will require a lot of changes on the security layer, complexify the configuration of argo, and require major changes in how the application-controller watches cluster. 

I think to achieve a level of multi-tenancy on the configuration layers of argo (everything but Application/ApplicationSet), it is a better solution to deploy dedicated instances in their own namespaces for each tenant. Perhaps some components such as the redis cache could be shared between instances. 

The team that deploys the argoCD instances can still manage the "global" configurations, while giving the "tenant owners" API access through RBAC to configure their resources for the instance they own.

Reply:
I'm not sure that having per-tenant Argo CD instances solves the problem of permitting self-service, GitOps-based cluster credential management. I can't grant tenant-2 GitOps management of secrets in the tenant-2 Argo CD namespace, because secrets in that namespace include things that should only be manageable by Argo CD admins.

Reply:
> I can't grant tenant-2 GitOps management of secrets in the tenant-2 Argo CD namespace, because secrets in that namespace include things that should only be manageable by Argo CD admins.

I agree, but with that deployment pattern, the problem statement is different. You only need to discover configs from another namespace, you do not need to have any concept of multi-tenancy between the config namespace and which applications can use them.

It can even be implemented today with a simple controller, deployed in argo namespace, that would watch for argo resources in another namespace and copy them in the local namespace. However, I think it would make sense to have that has a built-in feature in argo.

Reply:
Fair, I think it's probably time to set that long-term direction. And then if we can implement it incrementally, that'll be great.